### PR TITLE
[Support] Add parallel_for support to run a loop in parallel

### PR DESCRIPTION
**Pull request by jcf94**
As has mentioned in https://github.com/apache/incubator-tvm/pull/5962.
We would like a runtime implementation of parallel loop to speed up some thread safe loops.

Take some API reference from https://docs.microsoft.com/en-us/cpp/parallel/concrt/parallel-algorithms?view=vs-2019#parallel_for. I've tried to add an implementation of parallel_for_each as well, but finding that the const reference style of tvm::Array seems not very compatible with this API.

cc @tqchen @merrymercy @FrozenGene @junrushao1994

### CODE REVIEW

**Comment by jroesch** [comment]
cc @tkonolige

**Comment by tkonolige** [comment]
I think I'm a little late to this discussion, but what is the reason for having our own thread pool/parallel_for implementation? OpenMP is already an optional dependency, we could use it when available.

I can understand that having our own implementation means we don't have to depend on another library, but it also means we need to maintain it and add features as we need them (though this doesn't seem like much code).

**Comment by tqchen** [comment]
@jcf94 it would be great if we can simplify the parallel for implementation, e.g. std::thread has pretty low launching overhead, and we can likely drop the threadpool and start std::thread on each parallel_for round. As long as the function cost is not high, it should be a simpler implementation.

**Comment by jcf94** [comment]
> @jcf94 it would be great if we can simplify the parallel for implementation, e.g. std::thread has pretty low launching overhead, and we can likely drop the threadpool and start std::thread on each parallel_for round. As long as the function cost is not high, it should be a simpler implementation.

I've tried a simpler implementation using std::thread, see if this is better. : )

**Comment by tqchen** [comment]
Thanks @jcf94 please look into the CI issue. Also please do a benchmark to see if the new implementation will/will not affect perf, since it is only fine for larger functions and we might still need pool for very fine grained parallelism

File: src/support/parallel_for.cc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace tvm {
+namespace support {
+
+std::vector<std::vector<int>> rr_partitioner(int begin, int end, int step, int num_threads) {
+  int total_task_count = (end - begin) / step;
+  CHECK_GT(total_task_count, 0) << "Infinite loop condition, check the input value of "
+                                << "`begin`, `end`, `step`.";
+
+  std::vector<std::vector<int>> ret(num_threads);
```
Comment (merrymercy) [inline src/support/parallel_for.cc:41]: Add `std::vector::reserve` for all push_backs below when you know the length in advance.

File: src/support/parallel_for.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * should work the same as:
+ *   parallel_for(0, 10, [](int index) {
+ *     std::cout << index << "\n";
+ *   });
+ * \param begin The start index of this parallel loop(inclusive).
+ * \param end The end index of this parallel loop(exclusive).
+ * \param f The task function to be excuted. Assert to take an int index as input with no output.
+ * \param step The traversal step to the index.
+ * \param partitioner A partition function to split tasks to different threads. Use Round-robin
+ * partitioner in default.
```
Comment (merrymercy) [inline src/support/parallel_for.h:63]:
```suggestion
 * partitioner by default.
```

File: src/support/parallel_for.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * \param step The traversal step to the index.
+ * \param num_threads The number of threads(the number of tasks to be partitioned to).
+ * \return A list with `num_threads` elements, and each is a list of integers indicating the loop
+ * indexes for the corresponding thread to process.
+ */
+std::vector<std::vector<int>> rr_partitioner(int begin, int end, int step, int num_threads);
+
+/*!
+ * \brief A runtime api provided to run the task function in parallel.
+ * e.g. A for loop:
```
Comment (merrymercy) [inline src/support/parallel_for.h:50]: This is not a good example. We should mention the order of execution is not guaranteed.

Comment (jcf94) [reply src/support/parallel_for.h:50]: Updated.

File: src/support/parallel_for.cc [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      ret.push_back(std::vector<int>());
+    }
+    ret[thread].push_back(begin);
+  }
+  return ret;
+}
+
+void parallel_for(int begin, int end, const std::function<void(int)>& f, int step,
+                  const PartitionerFuncType partitioner) {
+  int default_num_threads = std::thread::hardware_concurrency();
```
Comment (FrozenGene) [inline src/support/parallel_for.cc:53]: Which condition we will use `parallel_for`? I want to make sure whether we will get the value like tvm thread pool (like for x86, we will ignore hyper thread.)

File: src/support/parallel_for.cc [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        });
+    res_vec.emplace_back(task.get_future());
+    threads.emplace_back(std::move(task), run_partition, f);
+  }
+
+  for (auto& thread : threads) {
+    thread.join();
+  }
+  try {
+    for (auto& i : res_vec) {
```
Comment (FrozenGene) [inline src/support/parallel_for.cc:75]: i think we don't change the value, so we could make it be `const auto&`?

Comment (FrozenGene) [inline src/support/parallel_for.cc:75]: i think we don't change the value, so we could make it be `const auto&`?

File: src/support/parallel_for.cc (hunk 4 above)
Comment (jcf94) [reply src/support/parallel_for.cc:53]: Currently we may use this in state SampleInitPopulation and InferBound on Array<State>, these're all thread independent.

File: src/support/parallel_for.cc (hunk 5 above)
Comment (jcf94) [reply src/support/parallel_for.cc:75]: Unfortunately the `thread.join()` and `future.get()` are all not const, they're not able to be call by a const reference.

**Comment by jcf94** [comment]
> Thanks @jcf94 please look into the CI issue. Also please do a benchmark to see if the new implementation will/will not affect perf, since it is only fine for larger functions and we might still need pool for very fine grained parallelism

Had some simple benchmark on the two implementations, the current one even works better in large loop size(since each threads' workload is pre-defined by the partitioner).
Currently Ansor's requirement on the `parallel_for` is just used on some thread independent tasks, this simple implementation is enough to work well. I'm not sure if some tasks with fine grained parallelism will benefit on a pool implementation.

File: src/support/parallel_for.cc (hunk 4 above)
Comment (FrozenGene) [reply src/support/parallel_for.cc:53]: If so, i think it is ok.

File: src/support/parallel_for.cc (hunk 5 above)
Comment (FrozenGene) [reply src/support/parallel_for.cc:75]: ok. I think we could use universal reference (https://zhuanlan.zhihu.com/p/25148592). However, nit-pick comment

**Comment by FrozenGene** [comment]
> @merrymercy @tqchen I'm thinking that do we need a interface to pass the `parapllel_for` to Python? In my understanding, the python multi-threading may have some performance issues caused by GIL.

for this, multi processing maybe enough...

File: tests/cpp/parallel_for_test.cc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "../../src/support/parallel_for.h"
```
Comment (merrymercy) [inline tests/cpp/parallel_for_test.cc:20]: Do not use the relative path. Add `parallel_for.h` to public header

**Comment by tqchen** [comment]
Thanks @jcf94 @merrymercy @jroesch @ @tkonolige !

**Pull request merged**